### ISSUE TITLE
Retiring the Mozilla Location Service

### ISSUE THREAD

**Issue by mostlygeek** [mozilla/ichnaea#2065]
The accuracy of [Mozilla Location Service ](https://location.services.mozilla.com/)(MLS) has steadily declined. With no plans to restart the stumbler program or increase investments to MLS we have made the decision to retire the service. 

In 2013, Mozilla launched MLS as an open service to provide geolocation lookups based on publicly observable radio signals. The service received community submissions of GPS data from the open source [MozStumbler](https://github.com/mozilla/MozStumbler) Android app. In 2019, Skyhook Holdings, Inc contacted Mozilla and alleged that MLS infringed a number of its patents. We reached an agreement with Skyhook that avoided litigation. This required us to [make changes to our MLS policies](https://blog.mozilla.org/services/2019/09/03/a-new-policy-for-mozilla-location-service/) and made it difficult for us to invest in and expand MLS. In early 2021, we [retired](https://discourse.mozilla.org/t/retiring-mozilla-stumbler/75206) the MozStumbler program. 

We are grateful for the contributions of the community to MLS to both the code and the dataset. To minimize disruptions and allow people time to make alternative arrangements, we have created a schedule that implements the retirement in stages. The retirement plan can be tracked in this issue. 

There will be five stages. 

1. As of today (Mar 13th, 2024) we will stop granting new API access keys. All pending applications will be rejected. 
2. On March 27th, 2024 we will stop accepting POST data submissions to the API. All submissions will receive a 403 response and the submitted data will be discarded. Additionally, we will stop publishing new exports of cell data for download. 
3. On April 10th, 2024 the cell data downloads will be deleted and will no longer be available. 
4. On June 12th, 2024 third party API keys will be removed and the service will only be available for Mozilla’s use cases. 
5. On July 31st, 2024 this source repo (https://github.com/mozilla/ichnaea) will be archived. 

The source code for the MLS application, Mozilla Ichnaea, will remain available under the Apache License 2.0.

**Comment by heftig** [comment]
Firefox still uses MLS for `browser.region.network.url`; will that also move to Google Location Services?

**Comment by alexcottner** [comment]
> Firefox still uses MLS for `browser.region.network.url`; will that also move to Google Location Services?

This endpoint will be migrated to another service (classify-client) that will return the expected response.

**Comment by heftig** [comment]
Will downstream builds of Firefox need to obtain API keys for this new service?

**Comment by mar-v-in** [comment]
@rtsisyk

Do we have any insights on which patents are applicable in this context and which jurisdictions they apply in? As you might know Europe is generally far less into software related patents, so maybe the "some jurisdiction" could be very easy to find.

The other question would be if Mozilla is willing to handover the WiFi dataset to a new organization running an ichnaea server or if we'd have to start from scratch.

**Comment by DylanVanAssche** [comment]
Also the Bluetooth beacons would be useful. I know Mozilla did not publish the WiFi and Bluetooth beacons for privacy reasons, but handing them over would be super beneficial. We would be set back by a decade if we have start from scratch.

**Comment by badrihippo** [comment]
What's the plan for KaiOS devices? IIRC they rely on MLS for location. I hope mine continues to function!

**Comment by vfosnar** [comment]
@thestinger 

> We have a plan for dealing with this using hardware-based attestation able to support submitting data from any modern Android device with the security model intact and recent security patches.

This feels bad. Can't see this implemented in a open source/community friendly way.

> We've never understood the privacy concern with providing a map of networks that are publicly broadcasting and are meant to be static landmarks.
+1

> The people who care about this tend to care about their privacy and probably don't want to be submitting this kind of data during regular use of their devices, especially if the data is kept tied to the accounts submitting it for anti-abuse.

Hot take, I'm okay with that. Trying to fight anti-abuse when only collecting anonymous data is impossible. As long as I can trust the provider (clear privacy policy, legal limitation on the use of provided data) and a good anonymization process of the data (wifi endpoints only published after multiple accounts have seen them independently) and obviously if both backend and frontend are open source.

**Comment by badrihippo** [comment]
> Hot take, I'm okay with that. Trying to fight anti-abuse when only collecting anonymous data is impossible. I would data connected to my account as long as I can trust the provider (clear privacy policy, legal limitation on the use of provided data) and a good anonymization process of the data (wifi endpoints only published after multiple accounts have seen them independently) and obviously if both backend and clients are open source.

I'm thinking of this as an "OpenStreetMap of location data". When people contribute locations to OSM, they are by nature implying that they were in the area—and they're fine with that.

Of course, location data is very different because the sharing process will likely be automated, not submitted through a conscious process. I can't speak for everyone, but I personally would be happy to submit this data to a trusted organisation. To make things even better, perhaps the account data can be de-linked after a reasonably long time, when most of the anti-abuse use-case has expired.

**Comment by cookieguru** [comment]
> We're not aware of sources for Wi-Fi and Bluetooth data

[wigle.net](https://wigle.net/)

**Comment by oguz-ismail** [comment]
good
firefox next

**Comment by Goldmaster** [comment]
I am a little surprised as it helps improve accuracy when GPS is not very good. It doesn't help that the option to help map was removed from the Web browser rewrite and that the stumbler app didn't get updated or made compatible with modern android versions.

**Comment by cookieguru** [comment]
> > > We're not aware of sources for Wi-Fi and Bluetooth data
> > 
> > 
> > [wigle.net](https://wigle.net/)
> 
> Unfortunately, this isn't open data for any usage

That is incorrect, [the EULA](https://wigle.net/eula.html) very explicitly states that you have a "right to use the maps and access point database...solely for your personal, research or educational, non-commercial purposes"

> It's not clear what they would allow but it's non-commercial usage only which somewhat implies not making a service usable for anything including commercial usage.

Your misuse of the apostrophe makes it difficult to decipher the intent, but if you're using the data for commercial purposes then you by definition have the funds to license the data.  Whether or not it the license fees are feasible for your project can only be speculated.

It certainly sounds like you're forging ahead on [this path](https://xkcd.com/927/)

**Comment by maciejsszmigiero** [comment]
Note that [Geoclue](https://gitlab.freedesktop.org/geoclue/geoclue) has always allowed submitting data to MLS from systems where it can access a GNSS receiver.

It is opt-in, however, both for privacy reasons and because that's what our users expect.

**Comment by aembleton** [comment]
> my 3c: I try to be privacy focused but at the same time I do (try to) contribute to openstreetmap, which requires account. 

The data collection could even be build into something like Street Complete.  As you're using it to update OSM, it could listen for SSIDs and feed those back.  It's already using the GPS for Street Complete and so wouldn't hurt the battery life.

**Comment by mar-v-in** [comment]
- For users that are very privacy aware, I suggest to prefetch the cell towers in their region and not use accurate wifi-based location at all. Prefetched cell tower data would also be sufficient for faster GPS via local SUPL, which doesn't need high accuracy and can handle inaccuracies of tens of kilometers pretty fine.
- I doubt that protecting a location service with device attestation makes a lot of sense for two reasons:
  1. You want people that are able and willing to collect wifi locations using highly sensitive long distance hardware to do that and contribute. This hardware is very likely not compliant with any attestations, but can be an extremely good source.
  2. Even when device attestation is used, it is "trivial" to introduce wrong records, as we're talking about uploading data that was previously retrieved from public radio airspace. Setting up a device that appears as a hotspot with arbitrary strength and mac address is easy enough for people that are willing to abuse.
- Publishing the raw data (both raw data of submissions and raw locations of wifi hotspots) can be a serious privacy issue - entirely independent of what privacy laws might state. As you are concerned with abuse, please also consider the abuse risk of this data. For example, stalkers will be able to follow their victims even when they move to a different location for as long as they continue to use the same wifi hardware.
- Requiring an account for uploading data is not generally a bad idea, however this account setup should be as easy as possible (e.g. could be just generating a unique random id client side). If passive contributing (that is: uploading data about wifi networks when GPS is already in use anyway) is easy to activate (e.g. just ticking a box) this would allow users to contribute without any downsides. A unique random id would still be useful for abuse prevention, as continued and proven correct submissions from a unique id could be used to increase trust in its data - so the "account" would only be used to gain trust, not to ban bad actors (which is hard anyways, as they will likely be able to create a new identity).

**Comment by mar-v-in** [comment]
> Wigle supports lookups by SSID/MAC already:
>
> https://wigle.net/
>
> Therefore, it doesn't seem to be a valid reason to oppose publishing open data. Many people believe in gathering and publishing this as data in the public domain usable for any purpose. A service making the data available to everyone has a much higher chance of success than one hoarding it and building a business model around it. 

Someone else doing it this way doesn't mean it's a good idea ;) There likely are reasons why the database of Mozilla had many more contributors even if the Wigle database is more open. I personally would not want to contribute to a database that can be easily abused by people with malicious intent.

And I wasn't suggesting to make a business model out of it, nor to "hoard" it. Making data not open to everyone is not the same as keeping data private. Data could be provided to others under clear rules that prevent or make abuse sufficiently hard, while not impacting its usefulness. For example, a rule could be that a full dump of raw data is provided to researchers, if usage is monitored by an IRB and all copies of the data is removed within a year after the research was conducted. I'm not saying it's easy to come up with appropriate rules for this, but I would argue it would be worth it to prevent abuse.

<details><summary>Off-topic: on accusations</summary>
As I previously said, I feel sorry for what happened to you and you are open to provide detailed accusations and supporting material by email to me, e.g. by mail to admin at microg.org, or to some independent third-party. And of course these can lead to a ban of individuals from the project. I don't think it's fair to claim wrongdoings by me personally because of a conflict you seem to have with an unnamed contributor to my project. I'd also like to point out that the project only has a single "main developer" beside myself and I doubt you are referring to that person, so maybe you should downgrade your wording to "minor contributor". And to reiterate the obvious: I don't want any contributions from Nazis and will happily ban those, no matter how good or how much code they were to contribute. But I also don't think this is the venue to discuss any of this.</details>

**Comment by Sapiosenses** [comment]
> Publishing the raw data (both raw data of submissions and raw locations of wifi hotspots) can be a serious privacy issue - entirely independent of what privacy laws might state. As you are concerned with abuse, please also consider the abuse risk of this data. For example, stalkers will be able to follow their victims even when they move to a different location for as long as they continue to use the same wifi hardware.
> 
> Requiring an account for uploading data is not generally a bad idea, however this account setup should be as easy as possible (e.g. could be just generating a unique random id client side). If passive contributing (that is: uploading data about wifi networks when GPS is already in use anyway) is easy to activate (e.g. just ticking a box) this would allow users to contribute without any downsides. A unique random id would still be useful for abuse prevention, as continued and proven correct submissions from a unique id could be used to increase trust in its data - so the "account" would only be used to gain trust, not to ban bad actors (which is hard anyways, as they will likely be able to create a new identity).


I am not particularly comfy with the idea of HW attestation for contributions. And I agree about the potential for abuse of user-generated RF emitter/geocoordinate data but I think there are ways this could be minimized. No time to expand on that rn unfortunately, mostly wanted to post my thoughts on the collection/validation of data below.

Given that we already have UNLP backends that build RF emitter databases based on approximate coordinates using a concurrently available GPS fix, seems to me that there could be some automated ways to validate that data without requiring all these draconian submitter requirements that, as you mention, could also be intentionally poisoned.

Just add a function to the UNLP backend in GMSCore that if the device has a current GPS fix and the reported geocoordinates of an RF emitter seem grossly inaccurate, have a way to send an anonymized "vote" back to the data server that that emitter source seems untrustworthy at that moment. Only after a minimal additional number of subsequent "sightings" verified the estimated location with *their* concurrent GPS fix, would it return to "validated" status in the database.

**Comment by Djfe** [comment]
This honestly leaves me feeling devastated. I'm staring at the screen with a bit of a blank expression 😮‍💨
it doesn't come by surprise, no, but I was actually still contributing from time to time using the Tower Collector app (even though this felt limited without wifi scans). I was always seeing the broader appeal of providing this type of service as open (source) to counter Google, Wigle etc. who don't give a * about privacy and didn't give you geofencing or whatever.

Please find a way to not throw the data away thar we collected all these years, even if lots of stuff is outdated now. It would mean a lot honestly!
I hope my message find's you well. I hope you see that MLS is actually cared about still.
Also: software patents are a b-word and I root for Mozilla for also being part of the movement against it.
I'm actually wondering: was moving MLS to European soil where software patents don't exist never a legal option?

is there a possibility to move the data into a new project with a new owner if anyone would care to step up and you are confident about them not abusing the data? or is that impossible due to copyright and such a company/legal entity not being Mozilla? (user only allowed Mozilla to use the data probably?)

Big thank you to anyone working on this back in the day. I really had a great time on github contributing in my way and I always felt appreciated by the devs. that was so awesome. MLS was for 3-5 years a big part of my hobbies. I used to cartography my neighborhood, city etc. in this way to contribute.
This felt important to me growing up as a teenager and the project gave me firm beliefs into the way and power of open source. It was the project that taught me commands like git rebase even if it was only to improve the German translation of the MozStumbler app 😅

I'm having quite the nostalgic moment right now.

**Comment by Sapiosenses** [comment]
> We'll be implementing our own client and server software. We don't and won't use that.

This is a microG issue, my comments pertain to microG.

You can do whatever you want.


> That kind of anti-abuse mechanism seems prone to abuse itself. People can fake being many different users if there isn't some way of preventing that.

The likelihood of anyone doing that and and especially succeeding with it, is very low.

I often spend time in places where there are hundreds of detectable RF emitters operated by both private and public entities, individuals and organizations. It's literally impossible for a malicious party to compromise all of them, and if one or two were compromised either internally or via using their MAC associated with a bogus set of geocoordinates it would be trivial to see that those are outliers whose data should be thrown out, especially as many other "viewers" would be passing by on a daily basis, throwing that data into question. 

Asking for draconian requirements for a crowdsourced collection of data would be like wikipedia demanding full legal government ID for all submitters to their database. They have come up with ways to ensure that the resulting product is not grotesquely incorrect, without imposing such draconian requirements on their contributors and vastly decreasing the number of people willing to contribute.

I doubt any reasonable person would expect or demand that every single record in that db be 100% accurate all the time. Case in point is how long people have put up with the Mozilla db despite the fact it has become extremely inaccurate for probably at least 2 years now. (That's how long ago I stopped using it for that reason)

Neither is HW attestation going to guarantee that accuracy.

**Comment by Sapiosenses** [comment]
> > We'll be implementing our own client and server software. We don't and won't use that.
> 
> This is a microG issue, my comments pertain to microG.
> 
> You can do whatever you want, obviously.
> 
> > That kind of anti-abuse mechanism seems prone to abuse itself. People can fake being many different users if there isn't some way of preventing that.
> 
> The likelihood of anyone doing that and and especially succeeding with it, is very low.
> 
> I often spend time in places where there are hundreds of detectable RF emitters operated by both private and public entities, individuals and organizations. It's literally impossible for a malicious party to compromise all of them, and if one or two were compromised either internally or via using their MAC associated with a bogus set of geocoordinates it would be trivial to see that those are outliers whose data should be thrown out, especially as many other "viewers" would be passing by on a daily basis, throwing that data into question.
> 
> Asking for draconian requirements for a crowdsourced collection of data would be like wikipedia demanding full legal government ID for all submitters to their database. They have come up with ways to ensure that the resulting product is not grotesquely incorrect, without imposing such draconian requirements on their contributors and vastly decreasing the number of people willing to contribute.
> 
> I doubt any reasonable person would expect or demand that every single record in that db be 100% accurate all the time. Case in point is how long people have put up with the Mozilla db despite the fact it has become extremely inaccurate for probably at least 2 years now. (That's how long ago I stopped using it for that reason)
> 
> Neither is HW attestation going to guarantee that accuracy.

**Comment by FedericoCeratto** [comment]
There are ways to reduce the privacy leaks and uploads of fake data ranging from simpler things like submitting cryptographic hashes of sets of macaddrs to using homomorphic cryptography as in https://ieeexplore.ieee.org/document/9079655

Yet, without millions of users uploading datapoints regularly the quality of any dataset would quickly degrade.

**Comment by unchartedxx** [comment]
>     2. On March 27th, 2024 we will stop accepting POST data submissions to the API. All submissions will receive a 403 response and the submitted data will be discarded. Additionally, we will stop publishing new exports of cell data for download.
> 
>     3. On April 10th, 2024 the cell data downloads will be deleted and will no longer be available.
> 

Could Mozilla provide full *cell* data exports about all the cells in their database before closing everything down? Even the full exports right now contain only the cells visible in the latest 12 months.

**Comment by janc13** [comment]
Stupid question/proposal: why are we collecting (always outdated & hard to verify) WiFi MAC/BSSID locations, when WiFi APs could (in the future) be broadcasting their (approximate) location instead?  Does anybody here have access to whatever forum is setting WiFi standards?

Alternatively/additionally, it should be possible to use mDNS and/or UPnP to request a (home) router’s (approximate) location.  That requires you have access to the LAN/WLAN of course, but would already cover many situations where people are indoors with no GNSS available, and it could be implemented with a very simple firmware update (most home routers already have UPnP & mDNS support, so only an option to set the location would have to be added).

Both (when widely implemented) would help provide accurate enough locations for a lot of use cases, without any need for massive data collection…

I understand that both methods could be abused to broadcast an incorrect location, but that would usually be very obvious & easily ignored. (And current methods can be and are being abused too—including GNSS.)

**Comment by mar-v-in** [comment]
@janc13 There is 802.11mc / Wi-Fi RTT / FTM which would solve this and would even allow for location accuracy far beyond typical GPS (field tests show less than 1m offset with 95% probability). However until now, only a small set of access points and client devices have implemented it and even less are configured with accurate location information.

**Comment by janc13** [comment]
That doesn’t seem to broadcast a location (except relative to the AP, which is useless if you don’t know its location)?

Also, RTT should not be needed for this, as in most cases it would be enough to know the location of the AP itself (usually you would be within ~10-50m, which is good enough for most purposes), and it seems to me like RTT would probably reveal _too much_ precision too easily.  People don’t need to know in what cupboard your AP is installed.

**Comment by mar-v-in** [comment]
The measurement report element of 802.11mc also includes the subelements Location Configuration Information and Location Civic Report, which provide latitude and longitude of the AP, but can also extended details like address, apartment number, room number and even down to seat or desk number.

The RTT name is a little bit misleading, but the reason to measure RTT this accurate in first place was to allow locating, that's why location data is provided at the same time.

I disagree, that 10-50m accuracy would be good enough in all cases. It's of course good enough for weather reports and probably car navigation, but can already be a significant difference for indoor navigation, especially if multiple floors are involved. Also for ((semi-) automated) emergency calling, accuracy down to 1m would be very important to have. Think about cases, where due to a medical emergency you are unable to speak: if the information that can provided to emergency services (via [AML](https://eena.org/our-work/eena-special-focus/advanced-mobile-location/)) is "the person that had a heart-attack is in this 10 story apartment complex with more than one hundred flats", I'm sure you would prefer to have 1m accurate indoor location ;)

**Comment by Sapiosenses** [comment]
> @janc13 There is 802.11mc / Wi-Fi RTT / FTM which would solve this and would even allow for location accuracy far beyond typical GPS (field tests show less than 1m offset with 95% probability). However until now, only a small set of access points and client devices have implemented it and even less are configured with accurate location information.

I'm including a screenshot of a settings page on my 2017 vintage Novatel/Inseego M7730L cellular hotspot which apparently implements this functionality. The latest generation of this device also has this feature and probably all the intervening models.

That said, I still prefer the idea of public collection of RF emitter data rather than relying on those who choose to enable this sort of functionality in their RF gear and do so in a non-malicious way.

![Novatel-Inseego_7730L_screenshot_GPS_feature](https://github.com/mozilla/ichnaea/assets/12429261/fb68714e-f966-4e91-ad8b-4e0728e65b0c)

**Comment by mar-v-in** [comment]
@Sapiosenses this looks as if it was a vendor specific custom protocol. microG implements several of those that are available in wifi networks in public transport

**Comment by alexcottner** [comment]
> Could Mozilla provide full cell data exports about all the cells in their database before closing everything down?

Great suggestion, we took a full dump of the cell tower data and have made it available for download [here](https://d2koia3g127518.cloudfront.net/final_export/MLS-full-cell-export-final.csv.gz).

**Comment by leag1234** [comment]
e Foundation (based in the EU; behind the /e/OS project), like other projects, is obviously looking for alternatives to MLS as well. We are ready to join and help credible initiatives in the field. 

Also there are remaning sub-grants up to 50K€  available from the MobiFree project that we have initiated (https://mobifree.org/). This would probably nicely fit with the development of (a part of) a new Location Service.

Also, e Foundation is ready to host such service.

Feel free to contact me about it.

**Comment by LucasMZReal** [comment]
Whatever service "we" come up with, I believe should have some properties:

* Cryptographic privacy (hashing for example, SSID + MAC, can help with making SSIDs not possible to easily look for in a public database, but findable by people who already know the details and just wanna find the location)
* Resistant to patents, as Mozilla has experienced, patents can sabotage the project (federation, decentralization can help here, I see an issue where independent data could be a problem, as people would contribute to different sources, and they wouldn't be merged, and none of them would be accurate or have enough data)
* Easily contributable
* Contributions should be bundled, this is a client thing, but submissions should be bundled IMO by default for every 7 days, so that a current location can't be revealed, and tracking is not possible, the more data there is around an area, the more it could update.
* Some resistance to profiling; It shouldn't be possible to check areas with more data and found out a person in a certain house, that goes to a certain school, goes to a certain clinic, is contributing to a low-contributor area. It's easy to connect dots that way.

**Comment by FedericoCeratto** [comment]
@LucasMZReal Indeed. As mentioned before there are various cryptographic methods to protect the SSIDs and macaddrs and prevent users to have their devices and traveling habits exposed. https://ieeexplore.ieee.org/document/9079655 is an example. Bundling a whole dataset might be impractical, but fetching and caching large areas e.g. a whole city or region should be easy.
Ideally the server-side component should rely on open data and not encourage user lock-in, so that multiple organizations could run the location service together.

**Comment by LucasMZReal** [comment]
Point is, I'm not contributing to a project that ignores privacy because "someone else is already doing it". These services need contributors, and I'm not about to be one of them for a service that's just as problematic.

And especially not to a project that won't help others, and undermine others.

**Comment by mar-v-in** [comment]
> long history of location data leaks to apps

I'm not aware of any location data leaking from microG. Mind opening an [issue](https://github.com/microg/GmsCore/issues) so it can be fixed?

> tying network-based location to Google service compatibility

microG already provides a system network location provider, so it's not tied to using the Google services, but can also be used from the AOSP network location provider interface (for apps that support that). It's also a [separate Gradle module](https://github.com/microg/GmsCore/tree/master/play-services-location/core/provider), so you can easily build it independent of the rest of microG
 
Of course only providing that interface is not enough as most apps use Google's location API, that's why GrapheneOS essentially did the same as microG here and provided an alternative implementation of Google's AIDL interface (using binder redirection instead of signature spoofing, both of which are techniques to make an app call a service that is not the one their code intended to call).

> If someone knows the SSID/MAC they can look it up if queries with one network are supported even with closed data. https://www.wigle.net/ and other services provide the ability

I don't think the database of Wigle can be used as a valid example to proof that data is generally accessible. I just tested against my personal wifi hotspot and there is no match in Wigle, but Apple, Google and Mozilla all have it in their database. Wigle's database is pretty large, but that's also because it has a ton of stale and outdated records. Mozilla Location Service had 50 times more submissions in half of the lifetime and still we get repeated complains by users about the quality of its database.

> SSID can be used to opt-out of mapping via `_nomap` but that's incredibly uncommon.

People want their devices to be able to geolocate themselves when at home, that doesn't mean they want every stranger that for some reason was able to discover their mac address to be able to find them. Also adding `_nomap` really sucks on the UX and many users are not even aware on how they can change their SSID.

> but downloading them for a whole region is still completely practical if it's stored efficiently

What is region in your mind here? It is estimated that in the western world, we have more than on wifi cell on average per person (some cells are shared, but many access points nowadays have at least 2 cells, one for 2.4GHz and one 5GHz). For the metropolitan region of NYC, that would mean more than 20M wifi cells. Then if you store their address (6 byte), latitude (4 byte), longitude (4 byte), altitude (2 byte) and signal strength modifier (1 byte) - which is the bare minimum of what you need for accurate indoor locating - you end up with more than 300MB of raw data for a single large region.

Now I do recognize that for performance and privacy reasons, it does make a lot of sense to have some data on the device. I personally see value in Apple's approach: When Apple devices request the Apple servers for a location based on wifi networks and the location can be resolved (i.e. the request is either for public hotspots or for at least three non-open access points), they provide the location data of the requested access points as well as the closest 100 access points nearby. This allows the Apple devices to cache that data locally and handle further location updates to some degree (e.g. when the user walks a little bit and therefor signal strength of cells changes and/or new cells show up) completely offline, while still only storing a small amount of data locally.

<hr>

I see there are some different approaches being discussed here and I think that makes a lot of sense. However I'd certainly wish for us to agree on a single common design and shared data that's being used by all of us - in the interest of having good data quality and the best service for all of our users and the users of other free software systems that require a replacement of Mozilla Location Service - I particularly also want to always include geoclue, which may not be as vocal here, but has a significant userbase. We really should make sure we don't end up in an XKCD#927 like situation.

And if financial resources are the issue here, I'm certainly willing to throw some money in the pot for GrapheneOS (or anyone else) to hire someone that works on providing this common solution for all of us.

**Comment by maciejsszmigiero** [comment]
The main problem here is securing funding for running the service and also for possible future R&D efforts.

If there was enough funding then it would be enough to simply keep running this software (_ichnaea_) on rented infrastructure to at least keep the service running past June.

This doesn't exclude further improvements in the future, if resources permit.

**Comment by mar-v-in** [comment]
> The main problem here is securing funding for running the service and also for possible future R&D efforts.

From what I read here and also heard from other sources, funding doesn't seem to be the main issue. /e/ offered to help with the hosting and for R&D, there is likely funds available through [NGI Mobifree](https://nlnet.nl/mobifree) or some other public funding source.

The main issues seem to be:
- What can legally be done? Patents seem to be an issue when running Ichnaea. This may lead to the requirement to ensure fully non-commercial operations or other restrictions, which some might be unwilling to.
- What do we want to do? Just running a new instance of Ichnaea would be easy, but also would result in insufficient data quality if we can't all agree on submitting to such a service.

**Comment by spaetz** [comment]
Just a bystander, but may I suggest that a few issues in here should be discussed in separate rooms/fora, in order to not derail from the core issue at hand. I don't think now (and this issue) is the right time and place to discuss implementation details of MLS v2.

I think the question about which patents are to be cared about and whoch jurisdictions are (most) vulnerable are pretty urgent ones to be clarified.

There is a foundation (thanks/e/) willing to help run it, so I would suggest one starts to setup an instance that can be used as a drop in replacement first, and goes from there.

**Comment by mar-v-in** [comment]
<details>
<summary>Off-Topic microG / GrapheneOS</summary>
> You must be aware you were leaking data to apps for years due to not implementing proxy AppOps enforcement.

I'm a bit surprised by you pulling out this issue that was fixed 4 years ago and only affected apps which did not target SDK 23 or higher (which at the time was already a requirement for apps uploaded to Play Store). Your phrasing made me thing there might be a new issue, but if this is what you were referring to, all good.

> The approach currently used in GrapheneOS properly enforces the permission model and attributes power usage / permission usage to the app since the app is sending the requests to the OS itself. We previously used the AppOps proxy / attribution APIs and found major issues with those along with it being overly complex and more prone to bugs.

Yes, as a privileged system component (being the OS itself) this is obviously easier. As microG is targeting to also run unprivileged / user-installed, this is unfortunately not that easy for us. I would see some benefits for splitting microG into multiple smaller pieces, which would make it easier and more reasonable to have some of them run privileged.

> I think it would be better to talk via direct message if you actually want to work anything out even if that simply means coming to a mutual understanding about a few things.

I wrote you on Matrix hoping to resolve this conflict outside this venue.

</details>

> How are you suggesting someone discovers it beyond physical access?

Examples could be:
- Software leaking mac addresses to unauthorized third-party software on the same device
- Software on nearby devices leaking mac addresses to third-party software running on adjacent devices

I'm just being realistic here, today it's pretty easy to get access to this identifier if you can run software on someone's device and that itself is also pretty easy as people use apps and websites with code these days, that have significant privileges on the systems they are executed.

> A drop-in replacement would require access to the current data.
> There's no way to simply continue the current service as is without access to the current data.

I think he was referring to an API drop-in, a server that implements the same API as MLS, so that systems that currently use MLS can switch easily. Even if the data in such a service is lower quality and/or it's only used for cell tower data and the IP-based fallback, it's probably better than nothing.

And of course it is possible to spin up the Ichnaea server software on a new machine without any of the wifi data and hope for users to contribute to it. It might not be the best approach technologically or when we look at *our users* privacy, but it might be the best approach when also considering the privacy of *people whose data is collected* and given the legal situation, as there seems to be consensus within privacy lawyers that wifi mac addresses are to be considered PII under GDPR, which likely makes sharing of this as open data illegal.

**Comment by alexcottner** [comment]
Hi everyone, 

This is a polite reminder that this is our professional working environment as much as it is our issue tracker.  I'm asking to keep comments focused on this issue and refrain from discussing projects outside of MLS or this repo.  I also encourage you to review our [Community Participation](https://www.mozilla.org/en-US/about/governance/policies/participation/) guidelines.   

Thank you.

**Comment by leag1234** [comment]
> @leag1234
> 
>     Again you and your toxic attitude are attacking us.
> 
> I have archives showing you falsely claiming GrapheneOS isn't open source and engaging in harassment targeting me including trying to portray me as crazy.
> 
>     I won't comment more: you are well known for your paranoia and for aggressing many projects, including us. And as I've told you before when you threatened to sue us for such and such overrealistic reasons (and this is endless and no-constructive discussions), we just prefer to stay away from you.
> 
> This is an example of you engaging in name calling and pushing fabricated claims about me as you have done many times. You have targeted other projects including DivestOS in the same way, and the DivestOS developer is being targeted with harassment too.
> 
>     Regarding e Foundation and you defamating claims, the open source community and people will judge by themselves.
> 
> The open source community at large recognizes that you're grifters. Engaging in Kiwi Farms style harassment targeting privacy and security researchers including at both DivestOS and GrapheneOS demonstrates who you are.

As usual... 

I will be interested in having a look at your "archives" and to (not)happy to discuss all this somewhere else, because this is about MLS.

**Comment by alexcottner** [comment]
Update - submission of data has now been disabled, completing stage 2 of the retirement process. Clients trying to submit data will see a 403 response with the below content.

```
{
  "error": {
    "errors": [
      {
        "domain": "global",
        "reason": "Forbidden",
        "message": "Forbidden"
      }
    ],
    "code": 403,
    "message": "Forbidden"
  }
}
```

**Comment by alexcottner** [comment]
> Will downstream builds of Firefox need to obtain API keys for this new service?

@heftig - We'd like to implement a simple plan that will allow downstream builds of firefox to self-identify with the classify-client service. Please see [this github issue](https://github.com/mozilla/classify-client/issues/151) for details.

**Comment by alexcottner** [comment]
Step 3 is now complete. The downloads page has been updated leaving only the final cell export listed.

**Comment by alexcottner** [comment]
Step 4, disabling of third party API keys, begins today. Requests to MLS will be receiving a 4xx error response, depending on the endpoint called (ex: `v1/geolocate` requests will receive 404). This shift will happen gradually over the next few days with the full migration targeted for Friday.
